daemon/image_service.go

```diff
@@ -0,0 +1,70 @@
+package daemon
+
+import (
+	"context"
+	"io"
+
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/reference"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/backend"
+	"github.com/docker/docker/api/types/filters"
+	imagetype "github.com/docker/docker/api/types/image"
+	"github.com/docker/docker/api/types/registry"
+	"github.com/docker/docker/builder"
+	"github.com/docker/docker/container"
+	"github.com/docker/docker/daemon/images"
+	"github.com/docker/docker/image"
+	"github.com/docker/docker/layer"
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+type ImageService interface {
```

Comment:
There's already a number of existing interfaces that cover this; wondering if we should (re)use those. Bit concerned about introducing a new interface that's really large 🤔

Reply:
To avoid such a big interface, and layers-related functions which seems to be only required for classic builder (?), I had in mind we could introduce interface on a per-command basis, typically to implement `image ls` we would introduce `ImageLister` interface with only `Images` func. Many of such interfaces could be implemented by a single component, but from daemon point of view, each feature would have a single interface defined.

Reply:
So yeah, there are other parts in the code that define only the subset of things they need from the image service.
<img width="579" alt="Screenshot 2022-06-28 at 15 24 48" src="https://user-images.githubusercontent.com/99933/176189889-247e7ac0-8830-4a48-97bd-543e4c78e321.png">

And then there is the daemon that pretty much uses all of these methods. The interface is indeed large but that's because, over the years, the image service became large. With this interface in one place it's easier to see what methods don't need to be in there: the layers methods for example, most of them are only passing the call down to the layer store.

Having a huge ugly interface can help us refactor the code and make it smaller.

Reply:
In other words, this huge interface shines a light on the size of the image service and will help us split it into more meaningful chunks

Reply:
True, although I think "image-service" was just a "let's give it a name"; basically it comes down to "whatever things were added to the daemon" and the "service" was only to, well, have a split between the API and daemon.

There are various interfaces for parts of the code that expect specific functions from the daemon (I think those are the ones you found above)
https://github.com/moby/moby/blob/1ba54a5fd0ba293db3bea46cd67604b593f2048b/image/tarexport/tarexport.go#L34-L37

https://github.com/moby/moby/blob/08615395719ca310693f92f7bc33c7c3ffe0ab09/api/server/router/image/backend.go#L17-L21


Those are not as much split-up as I'd love them to see, but at least it could be a start; and perhaps some of the functions that are not in the interface should/could be defined as a temporary interface if we need to.

Reply:
There's also https://github.com/moby/moby/pull/43697, which reduces some interfaces (oh! never mind, that's for containers, not images, but could still be OK to have - Tonis wondered if we needed it as (part of) that would be removed in the migration, OTOH, having a smaller interface to start with could still be useful perhaps).